califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
+
+            if (cryId > fTotalCrystals && cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryId <= fTotalCrystals && cryEnergy > fSaturation)
+                saturatedCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+
+        /* -------- Simulation data ------- */
+        else
+        {
+
+            if (cryEnergy >= fGammaClusterThreshold && cryEnergy < fProtonClusterThreshold)
+                gammaCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+
+            if (cryEnergy >= fProtonClusterThreshold)
+                protonCandidatesVec.push_back(dynamic_cast<R3BCalifaCrystalCalData*>(fCrystalCalData->At(i)));
+        }
+    }
+
+    /* ----- Remove duplicate entries in gamma clusters ----- */
+    for (int k = 0; k < gammaCandidatesVec.size(); k++)
+    {
+        for (int s = 0; s < gammaCandidatesVec.size(); s++)
+        {
+
+            if (gammaCandidatesVec.at(k)->GetCrystalId() == (gammaCandidatesVec.at(s)->GetCrystalId() - fTotalCrystals))
+                gammaCandidatesVec.erase(gammaCandidatesVec.begin() + s);
+        }
+    }
+
+    // Sort all vectors by energy
+    std::sort(gammaCandidatesVec.begin(), gammaCandidatesVec.end(), compareByEnergy);
+    std::sort(protonCandidatesVec.begin(), protonCandidatesVec.end(), compareByEnergy);
+
+    TVector3 mother_angles, angles;
+    Double_t fRandTheta, fRandPhi;
+
+    while (protonCandidatesVec.size())
+    {
+
+        Int_t motherId = protonCandidatesVec.at(0)->GetCrystalId();
+
+        califa_candidate cluster = { motherId, vector<Int_t>(), 0.0, 0.0, 0.0, 0.0, 0.0 };
+
+        if (motherId <= fTotalCrystals)
+            mother_angles = fCalifaGeo->GetAngles(motherId) - fCalifatoTargetPos;
+
+        else
+            mother_angles = fCalifaGeo->GetAngles(motherId - fTotalCrystals) - fCalifatoTargetPos;
+
+        if (fRand)
+        {
+
+            if (motherId > fTotalCrystals)
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1 - fTotalCrystals]->GetRandom2(
+                    fRandPhi, fRandTheta);
+
+            else
+                fAngularDistributions[protonCandidatesVec.at(0)->GetCrystalId() - 1]->GetRandom2(fRandPhi, fRandTheta);
+
+            cluster.theta = TMath::DegToRad() * fRandTheta;
+            cluster.phi = TMath::DegToRad() * fRandPhi;
+        }
+
+        else
+        {
+
+            cluster.theta = mother_angles.Theta();
+            cluster.phi = mother_angles.Phi();
+        }
+
+        cluster.time = protonCandidatesVec.at(0)->GetTime();
+
+        addCrystal2Cluster(&cluster, protonCandidatesVec.at(0), "proton", &usedCrystals, fTotalCrystals);
+
+        for (Int_t j = 0; j < allCrystalVec.size(); j++)
+        {
+
+            Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId();
```

Comment:
Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId()>fTotalCrystals?allCrystalVec.at(j)->GetCrystalId()-fTotalCrystals:allCrystalVec.at(j)->GetCrystalId();
Then life will be easier

Reply:
That won't work. You are killing the double reading with that statement.

Reply:
Okay then how about like that?
Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId();
string range = "gamma";
if(thisCryId > fTotalCrystals)
{
thisCrysId -= fTotalCrystals;
range = "proton";
}
Because other parts are the same for both ranges.